### ISSUE TITLE
Is this a 'finished' project that works?

### ISSUE THREAD

**Issue by gavspav** [Reinissance/MobMuPlatLink#1]
MobMuPlatLink - wow! Before I go through the pain of trying to teach myself how to build this you couldn't let me know if it works could you? I'm not quite sure if it is work in development or not.
Also is this the most recent version of MobMuPlat and if not can I integrate the link parts in that?
I don't know what I'm doing but I'm very stubborn so I'll probably work it out.

**Comment by Reinissance** [comment]
Hi, to me it works so far: I also included that You can load multiple patches of pd or mmp at once...

After you were able to build the project properly (I think you have to change the bundle identifier, maybe some other things too...), locate the MobMuPlatLink.app in Finder and cmd-click to show package content of the app bundle to get the Info.plist file:
Then You have to register yourself with your apple id at the [audiobus developer center](https://developer.audiob.us/) to create a [temporary registration](https://developer.audiob.us/temporary-registration) were you drop the Info.plist file. Then make sure the Settings look like this (except for the bundle identifier):

![bildschirmfoto 2019-03-05 um 14 49 53](https://user-images.githubusercontent.com/2316607/53810025-55c19c80-3f56-11e9-9158-b1effaee78f6.png)
![bildschirmfoto 2019-03-05 um 14 50 06](https://user-images.githubusercontent.com/2316607/53810047-5f4b0480-3f56-11e9-9426-060584793a16.png)

...and submit. Copy the code you get into the MMPViewController.m file, line 544, replacing the string that looks the same.
After that rebuild and everything it should be working for 14 days (allthough I have to rebuild after 7 days, because I have no payed developer account), after that time you have to repeat the steps above.

Thus You should be able to enjoy MobMuPlat with AbletonLink...

One more thing: if no other apps or members are joined to the Link-Session, loading new patches change the BPM of the session to the default setting of the new patch. You can prevent this behavior by embeding MobMuPlatLink into another audiobus host like AUM for example, which has Link enabled.

Good luck!

**Comment by gavspav** [comment]
Hmm I've followed your instructions I think.
I've built it and registered with Audiobus. It appears to run but there's just a black screen. Nothing happening.
Here's a screenshot of the only thing I couldsee in the debugger
Any ideas?

![screen shot 2019-03-06 at 12 33 34](https://user-images.githubusercontent.com/5639214/53881808-819f5980-400c-11e9-9485-855eea3603e9.png)

**Comment by gavspav** [comment]
Found some more debug info

2019-03-06 14:41:06.097156+0000 MobMuPlat[1438:539069] +[NSString(OSCStringAdditions) load]
2019-03-06 14:41:06.235359+0000 MobMuPlat[1438:539069] [MC] System group container for systemgroup.com.apple.configurationprofiles path is /private/var/containers/Shared/SystemGroup/systemgroup.com.apple.configurationprofiles

2019-03-06 14:41:06.236245+0000 MobMuPlat[1438:539069] [MC] Reading from public effective user settings.
2019-03-06 14:41:07.139973+0000 MobMuPlat[1438:539152] [avas] AVAudioSessionPortImpl.mm:56:ValidateRequiredFields: Unknown selected data source for Port Speaker (type: Speaker)
2019-03-06 14:41:07.140140+0000 MobMuPlat[1438:539152] Route changed.
2019-03-06 14:41:07.146500+0000 MobMuPlat[1438:539069] This Audiobus API Key is invalid for this app (code 4)
ABFilterPort.m:136: You must provide an audio unit
(lldb)

**Comment by gavspav** [comment]
Hmm can't seem to get the API Key to take.
Would that cause the black screen issue anyway?
I'm slashing away in the dark a bit. Gonna do something else for a bit!

**Comment by Reinissance** [comment]
Ahm... have to admit I‘m also not actually knowing what I’m doing until I get what I want...
I‘m not at my laptop now, but one thing I might have forgotten: In the Info.plist file there is a manufacturer value „rIni“ that I I remember I have changed to this. You can try to change it to something like „gAvs“ or so, but you have to do it in Xcode in the Info tab of the project I think. Then rebuild, get another Audiobus Key and try again...

**Comment by gavspav** [comment]
Ha! You mean the manufacturer name in the Audio Component? I changed this in info.plist 
Is there somewhere I can change it in Xcode. I'll have a look.
Also I notice there are 2 repos - MobMuPlat and MobMuPlatLink - is this one the most up to date? I think I might have cloned and built the other one (MobMuPlat)

**Comment by Reinissance** [comment]
Link is the one Most update... I don’t know if the other is working properly, I currently using this one.

**Comment by Reinissance** [comment]
In Xcode when you select the project near the general tab you find the info tab. I think that’s actually the plist file you also simply can edit from the file inspector too.

**Comment by gavspav** [comment]
Ok I've got the temp API key sorted, got the app built and it seems to be runnning but I can't see anything on the screen! Any ideas? 

Here's the debug info:

2019-03-07 15:46:47.525649+0000 MobMuPlatLink[1889:658488] +[NSString(OSCStringAdditions) load]
2019-03-07 15:46:47.645313+0000 MobMuPlatLink[1889:658488] [MC] System group container for systemgroup.com.apple.configurationprofiles path is /private/var/containers/Shared/SystemGroup/systemgroup.com.apple.configurationprofiles
2019-03-07 15:46:47.645548+0000 MobMuPlatLink[1889:658488] [MC] Reading from public effective user settings.
2019-03-07 15:46:48.680816+0000 MobMuPlatLink[1889:658536] [avas] AVAudioSessionPortImpl.mm:56:ValidateRequiredFields: Unknown selected data source for Port Speaker (type: Speaker)
2019-03-07 15:46:48.681271+0000 MobMuPlatLink[1889:658536] Route changed.
2019-03-07 15:46:48.751963+0000 MobMuPlatLink[1889:658488] Temporary Audiobus API Key: 14 days remaining
2019-03-07 15:46:48.752465+0000 MobMuPlatLink[1889:658488] CFSocketSetAddress listen failure: 102
2019-03-07 15:46:48.752571+0000 MobMuPlatLink[1889:658488] CFSocketSetAddress listen failure: 102
2019-03-07 15:46:48.752682+0000 MobMuPlatLink[1889:658488] Local endpoint is MobMuPlatLink (BDE7A7A4-2C2F-4796-BE0A-0CCDCE637F9E), port 54357
ABSenderPort.m:172: You must declare this audio component within your Info.plist's AudioComponents list

**Comment by Reinissance** [comment]
Okay, I will do boss is set up properly.I sometimes had a similar situation which I was able to solve by simply restarting the app.I mean close it, and start again, did you try that?

**Comment by Reinissance** [comment]
Ups, “Audiobus seems to be...” - Siri doesn’t seem to know this...

**Comment by Reinissance** [comment]
Because the ports are declared in the plist file, if you followed the steps above... And if I close it and restart from the phone itself it works usually, and does not happen again. Only if you have an unstable WiFi connection the app sometimes crashes, otherwise it actually works great.

**Comment by gavspav** [comment]
Hmm no restarting it is the same deal. Just a black screen.

Could audiobus or link or Audio resource issues cause that?

**Comment by gavspav** [comment]
Hmm just tried it on my phone and it works (no api key yet)!
Not sure why it wouldn’t work on the iPad
Have you got a link demo patch I can test it with?

**Comment by Reinissance** [comment]
In the mmp- examples there is a link example, where you can see at least the link running, although with no audio yet... I’ll send something when I’m at home

**Comment by Reinissance** [comment]
Well, I dont have an iPad, but in the simulator it works...
On Your phone as well, right? No clue about that ipad, - which device is it?
A patch is attached...
[planeTrance.zip](https://github.com/Reinissance/MobMuPlatLink/files/2942878/planeTrance.zip)

**Comment by gavspav** [comment]
Oops I spoke to soon.
I was opening the wrong MobMuPlat on my phone! Doh.
Can’t get any of them to work on any device or simulator.
Just a black screen!

**Comment by gavspav** [comment]
{"app_name":"MobMuPlatLink","timestamp":"2019-03-08 08:44:21.68
+0000","app_version":"1.81","slice_uuid":"f05fe2a4-be7f-3ecd-9fa7-9d1d06c389a5","adam_id":0,"build_version":"1.81.1","bundleID":"gav.reinissance.MobMuPlatLink","share_with_app_devs":false,"is_first_party":false,"bug_type":"109","os_version":"iPhone
OS 11.4.1
(15G77)","incident_id":"25D22452-A791-4240-A3EE-3E70846EA53B","name":"MobMuPlatLink"}

Incident Identifier: 25D22452-A791-4240-A3EE-3E70846EA53B

CrashReporter Key:   45d42af7f21ba9ff389617d2b3887202993b66f0

Hardware Model:      iPhone8,4

Process:             MobMuPlatLink [906]

Path:
/private/var/containers/Bundle/Application/5A5A0868-BCCE-4CFB-8B2C-6629FCE6F8CE/MobMuPlatLink.app/MobMuPlatLink

Identifier:          gav.reinissance.MobMuPlatLink

Version:             1.81.1 (1.81)

Code Type:           ARM-64 (Native)

Role:                Foreground

Parent Process:      launchd [1]

Coalition:           gav.reinissance.MobMuPlatLink [641]



Date/Time:           2019-03-08 08:44:21.6041 +0000

Launch Time:         2019-03-08 08:44:20.5160 +0000

OS Version:          iPhone OS 11.4.1 (15G77)

Baseband Version:    6.80.00

Report Version:      104


Exception Type:  EXC_CRASH (SIGABRT)

Exception Codes: 0x0000000000000000, 0x0000000000000000

Exception Note:  EXC_CORPSE_NOTIFY

Triggered by Thread:  0


Application Specific Information:

abort() called


Filtered syslog:

None found


Thread 0 name:  Dispatch queue: com.apple.main-thread

Thread 0 Crashed:

0   libsystem_kernel.dylib        0x00000001813c12e0 0x18139f000 + 140000

1   libsystem_pthread.dylib       0x0000000181562288 0x18155f000 + 12936

2   libsystem_c.dylib             0x000000018132fd0c 0x1812cd000 + 404748

3   MobMuPlatLink                 0x0000000102976908 0x10276c000 + 2140424

4   MobMuPlatLink                 0x0000000102971af4 0x10276c000 + 2120436

5   MobMuPlatLink                 0x000000010297120c 0x10276c000 + 2118156

6   MobMuPlatLink                 0x000000010277c86c 0x10276c000 + 67692

7   MobMuPlatLink                 0x00000001027795ac 0x10276c000 + 54700

8   MobMuPlatLink                 0x000000010277894c 0x10276c000 + 51532

9   MobMuPlatLink                 0x0000000102776c20 0x10276c000 + 44064

10  UIKit                         0x000000018b549dbc 0x18b503000 + 290236

11  UIKit                         0x000000018b5491c4 0x18b503000 + 287172

12  UIKit                         0x000000018b5165e0 0x18b503000 + 79328

13  UIKit                         0x000000018bb46b1c 0x18b503000 + 6568732

14  UIKit                         0x000000018b515dd0 0x18b503000 + 77264

15  UIKit                         0x000000018b515c6c 0x18b503000 + 76908

16  UIKit                         0x000000018b514afc 0x18b503000 + 72444

17  UIKit                         0x000000018c1aa84c 0x18b503000 + 13269068

18  UIKit                         0x000000018b5141ec 0x18b503000 + 70124

19  UIKit                         0x000000018bf8fac8 0x18b503000 + 11061960

20  UIKit                         0x000000018c0ddbf8 0x18b503000 + 12430328

21  UIKit                         0x000000018b513c0c 0x18b503000 + 68620

22  UIKit                         0x000000018b5135a8 0x18b503000 + 66984

23  UIKit                         0x000000018b5105e0 0x18b503000 + 54752

24  UIKit                         0x000000018b510330 0x18b503000 + 54064

25  FrontBoardServices            0x000000018413c470 0x184130000 + 50288

26  FrontBoardServices            0x0000000184144d6c 0x184130000 + 85356

27  libdispatch.dylib             0x000000018122ca60 0x18122b000 + 6752

28  libdispatch.dylib             0x0000000181234170 0x18122b000 + 37232

29  FrontBoardServices            0x0000000184170878 0x184130000 + 264312

30  FrontBoardServices            0x000000018417051c 0x184130000 + 263452

31  FrontBoardServices            0x0000000184170ab8 0x184130000 + 264888

32  CoreFoundation                0x00000001818e3404 0x1817f5000 + 975876

33  CoreFoundation                0x00000001818e2c2c 0x1817f5000 + 973868

34  CoreFoundation                0x00000001818e079c 0x1817f5000 + 964508

35  CoreFoundation                0x0000000181800da8 0x1817f5000 + 48552

36  GraphicsServices              0x00000001837e6020 0x1837db000 + 45088

37  UIKit                         0x000000018b820758 0x18b503000 + 3266392

38  MobMuPlatLink                 0x00000001027769cc 0x10276c000 + 43468

39  libdyld.dylib                 0x0000000181291fc0 0x181291000 + 4032

**Comment by gavspav** [comment]
Yes. Success. Finally got this working! 
I had to make 2 new audio components in info.plist for senderACD and filterACD
Or should I have just edited the ones that were there?
Looking back you may have been telling me I had to do this right from the start but I just thought I had to rename them. 
Anyway - seems to work now. Thanks for your help

**Comment by Reinissance** [comment]
So sorry for that.as I mentioned, I also am not exactly knowing what I’m doing, or what’s going on. so in that sense I appreciate your help!
Glad it works for you now!

**Comment by gavspav** [comment]
Well done for getting it to work. How did you do it? I mean what did you
have to do?
Have you contacted the guy who made MobMuPlat to see if he wants to put it
in the main project?
Gavin

**Comment by gavspav** [comment]
I spoke a bit soon there.
The link example does work in that the bmp changes in line with other
linked apps.
But the phase slider doesn't move, the count doesn't change and the
beattime is stuck on -484929.
It works on my mac with the editor open so I guess the link object isn't
built properly?
I do get this mach-o linker warning: ld: warning: directory not found for
option '-L../../LinkKit/lib'

**Comment by Reinissance** [comment]
OMG, there were indeed files missing... I just pushed a new commit. Will you please pull?

**Comment by Reinissance** [comment]
Yes, I wrote to the developer but as you see it wasn’t finished... maybe he will after we got through this.

**Comment by gavspav** [comment]
Ah thanks. I’ll try to find some time later to look at this. Bit busy over
the weekend. Hopefully it will ‘just work’ now!

**Comment by Reinissance** [comment]
In case not have a look in build phases if all required libraries are there (red ones indicate they’re not)

**Comment by Reinissance** [comment]
Maybe also for the frameworks, allthough without them it should not have been able to build I think

**Comment by gavspav** [comment]
I couldn't pull from git as I had made some changes to local files and it wouldn't let me. 
In any case I thought it would be useful if I started afresh and documented everything because then you can make sure the repository works for other people (if you want/have time).

So I'm using Xcode 10.1 on a 2011 15" MBP running High Sierra 10.13.6
I've actually got 2 iPads and an iPhone here running err 12.1,11.4.1 and 9.3.2

So I've cloned the repository into my Documents folder.
Next I've run git submodule update --init --recursive
Opened the Xcode Project

The first thing I notice is that there are several files missing (in red). On closer inspection these are all files which Xcode thinks are in ../MobMuPlatIos/MobMuPlatLink but in the repo they are in fact in ../MobMuPlatIos/MobMuPlat/ Files like MobMuPlatLink.entitlements, SettingsViewController etc. I'm not sure whether to move the files in the directory or change the links. Or if it matters which.

There are also a couple of frameworks not linking correctly ie in red. The first is libABLLink.a which I find in the MobMuPlat-iOS folder. The second is libz.dylib. I can't find this with the repo but I can find it on my hard drive in some kind of Developer/Ios folder. I have two copies of the library and they are the same so I cross my fingers it will work and link to that!

Next I change the Bundle Identifier and the signing team.
The Launch Images Source - LaunchImage is in red I can't work out how to change this.
It is saying the entitlements file could not be opened - I hope this will be sorted now I have updated the link to the file.

I'm going to try a first build....

I get an error in libpd s_file.c 'system' is unavailable: not available on iOS. After a search I think this is non critical and comment it out. //system(cmdbuf);

I build again and get an error about some missing files. I look again and realise that all the files Xcode thinks are in MobMuPlatLink are not there. Xcode doesn't know about the MobMuPlat folder (where the files are on my HD). So I just copy the files into the MobMuPlatLink folder. There are a couple of duplicated files - I keep the version in to MobMuPlatLink folder.

There are still some files missing in the graphics folder so I copy the LaunchImage.Launch image folder across and that fixes it.

Then I realise the entitlements file warning in the general information is because the file is incorrectly named in the CodeSigning Entitlements - it refers to MobMuPlatLink.entitlements but the file in the repo is called MobMuPlat.entitlements. I change this.

I build again and get ARC Semantic Issue - no visible @interface for Widget .......SceneViewController.m

I am a bit stuck now. I change the project settings to Legacy Build Settings and try to run the project in the simulator. I have two errors - Property autoLoad not found on object of type SettingsViewController and no visible @interface for SettingsViewCotroller.

Yup stuck!. There are still some steps which I know I have to do - like changing the Audio Component info.plist and registering the API but I haven't got that far yet. I'm going to have to ask for your help here????

**Comment by Reinissance** [comment]
Ok, sorry for not answering for so long, been busy...
I`m gonna make it as You did: clone for repositiory and get it running, documenting steps.
Meanwhile I just did this with the other repository, MobMuPlat, which worked fine: Just had to comment out the above mentioned line in libpd.xcodeproj/pure-data/stc/s_file.c line 281, then it works, - even no temporary key for audiobus was necessary.
Sorry for that, but I should have answered Your question above this way, but I didn`t know...
So first go ahead and grab the other one, until I get this rep running...

**Comment by gavspav** [comment]
Ah thanks for getting back - I thought you'd given up on me!
I've been trying to get the MobMuPlatLink one working on and off with no success! I've got and error saying:
 "No visible @interface for 'Widget' declares the selector 'replaceDollarZerosForGui:withInteger:'"

Xcode is confusing! And it's hard to follow other people's projects when you don't know what you're doing.!

I'm going to try and build the other repo now. I take it you meant the MobMuPlat on your github page and not the original non link one (which incidentally compiles and runs fine)

What's the difference between your two repo's anyway?

**Comment by gavspav** [comment]
I managed to get the first repo working - success!
I did have to do a few things to get it working - link the two libs and change the build settings to legacy - but not too much.
It won't work on the iphone se though because the Link button is hidden in the audio midi settings so you can't switch Link on!

**Comment by Reinissance** [comment]
Ok, now I think got it.
You can now clone, ubdate submodules as you described above, comment out the one line, change the bundle identifier, add another value as manufacturer to the audiocomponents in the plist file, get yourself a temporary audiobus key and this should be it.
Will you try and let me know?
Thanks for your help!!!

**Comment by gavspav** [comment]
The Mobmuplatlink repo?

**Comment by Reinissance** [comment]
Yes, this one.
For the other one, MobMuPlat, I pushed a commit, where the the Ableton Button now should be visible.

**Comment by gavspav** [comment]
Yes I can confirm that this Repo now works! The only steps I had to take were changing the Bundle Identifier, setting the signing Team, relinking the libABLLink and libz.dyib libariers and commenting out the system(cmdbuf) line. I didn't even edit the audio component stuff although this may stop things from working - there is audio though and Link works properly.

**Comment by gavspav** [comment]
Hey thanks for all your work on this.
The Button works as well.
I'm still a bit unsure as to which is the best version but they are both
working. Maybe delete one of them? It's a bit confusing!
Let me know if you need anything else testing,
Gavin
